vending-app/src/main/kotlin/com/google/android/finsky/DownloadManager.kt

```diff
@@ -0,0 +1,247 @@
+            val channel = NotificationChannel(CHANNEL_ID, name, importance).apply {
+                description = descriptionText
+            }
+            val notificationManager: NotificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
+            notificationManager.createNotificationChannel(channel)
+        }
+    }
+
+    private val executor by lazy {
+        ThreadPoolExecutor(
+            corePoolSize, maximumPoolSize, keepAliveTime, TimeUnit.SECONDS, LinkedBlockingQueue()
+        ) { r -> Thread(r).apply { name = "DownloadThread" } }
+    }
+
+    private fun initNotification(moduleName: String, packageName: String) {
+        val cancelIntent = Intent(CANCEL_ACTION).apply {
+            putExtra(KEY_MODULE_NAME, moduleName)
+        }
+        val cancelPendingIntent = PendingIntent.getBroadcast(
+            context, 0, cancelIntent, PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
+        )
+        val packageManager: PackageManager = context.packageManager
+        val applicationInfo = packageManager.getApplicationInfo(packageName, 0)
+        val appName = packageManager.getApplicationLabel(applicationInfo).toString()
+        val appIcon = packageManager.getApplicationIcon(applicationInfo)
+        val bitmap = if (appIcon is BitmapDrawable) {
+            appIcon.bitmap
+        } else {
+            val bitmapTemp = Bitmap.createBitmap(
+                appIcon.intrinsicWidth, appIcon.intrinsicHeight, Bitmap.Config.ARGB_8888
+            )
+            val canvas = Canvas(bitmapTemp)
+            appIcon.setBounds(0, 0, canvas.width, canvas.height)
+            appIcon.draw(canvas)
+            bitmapTemp
+        }
+
+        val notificationLayout = RemoteViews(context.packageName, R.layout.layout_download_notification)
+        notificationLayout.setTextViewText(
+            R.id.notification_title, context.getString(R.string.download_notification_attachment_file, appName)
+        )
+        notificationLayout.setTextViewText(
+            R.id.notification_text, context.getString(R.string.download_notification_tips)
+        )
+        notificationLayout.setProgressBar(R.id.progress_bar, 100, 0, false)
+        notificationLayout.setImageViewBitmap(R.id.app_icon, bitmap)
+        notificationLayout.setOnClickPendingIntent(R.id.cancel_button, cancelPendingIntent)
+
+        val notifyBuilder =
+            NotificationCompat.Builder(context, CHANNEL_ID).setSmallIcon(R.drawable.ic_app_foreground).setStyle(NotificationCompat.DecoratedCustomViewStyle())
+                .setCustomContentView(notificationLayout).setPriority(NotificationCompat.PRIORITY_LOW).setOngoing(true).setOnlyAlertOnce(true)
+                .setColor(ContextCompat.getColor(context, R.color.notification_color)).setColorized(true)
+        notifyBuilderMap[moduleName] = notifyBuilder
+        notificationLayoutMap[moduleName] = notificationLayout
```

Comment:
Any reason you use a custom notification layout instead of a default with [setContentTitle()](https://developer.android.com/reference/androidx/core/app/NotificationCompat.Builder#setContentTitle(java.lang.CharSequence)), [setContentText()](https://developer.android.com/reference/androidx/core/app/NotificationCompat.Builder#setContentText(java.lang.CharSequence)), [setProgress()](https://developer.android.com/reference/androidx/core/app/NotificationCompat.Builder#setProgress(int,int,boolean)), [setLargeIcon()](https://developer.android.com/reference/androidx/core/app/NotificationCompat.Builder#setLargeIcon(android.graphics.Bitmap)) and [addAction()](https://developer.android.com/reference/androidx/core/app/NotificationCompat.Builder#addAction(int,java.lang.CharSequence,android.app.PendingIntent))? That should cover the same content but not require a custom view with custom styling, thus better compatible with various android versions.